dali/pipeline/executor/executor2/exec_graph.h

```diff
@@ -0,0 +1,362 @@
+   *
+   * op A --- (output 0) ----- op B --------- ...
+   *    \                         /
+   *     ------- (output 1) -----(- op C --- ...
+   *                              \   |
+   *                               \  |
+   *                                \  \
+   *                                 \_release outptus (A)
+   */
+  tasking::SharedTask release_outputs_;
+
+  /** Creates a workspace at the pipeline's output.
+   *
+   * Output node's inputs are converted to Pipeline's outputs. The output workspace doesn't have
+   * any inputs.
+   */
+  CachedWorkspace CreateOutputWorkspace();
+  /** Creates a workspace suitable for running the operator `op`. */
+  CachedWorkspace CreateOpWorkspace();
+
+  WorkspaceCache workspace_cache_;
+
+  /** Moves to a new iteration. */
+  void NextIter() {
+    prev_task_ = std::move(main_task_);
+    release_outputs_.reset();
+  }
+
+  friend class ExecGraph;
+
+  /** Creates the main task.
+   *
+   * The graph is built in multiple phases. First, the main tasks are created for all operators.
+   * Then, data dependencies are added.
+   * Finally, auxiliary tasks (e.g. release_outputs_) are added. Creating all tasks for a node
+   * cannot be done in one go because release_outputs_ succeeds the main tasks of this node's
+   * consumers.
+   */
+  void CreateMainTask(std::shared_ptr<IterationData> iter, const WorkspaceParams &params);
+  /** Subscribes to the results of the precending tasks. */
+  void AddDataDeps();
+  /** Creates auxiliary tasks and applies concurrency constraints. */
+  void CreateAuxTasks();
+  std::optional<tasking::TaskFuture> Launch(tasking::Scheduler &sched);
+};
+
+/** The execution graph */
+class DLL_PUBLIC ExecGraph {
+ public:
+  std::list<ExecNode> &Nodes() {
+    return nodes_;
+  }
+
+  std::list<ExecEdge> &Edges() {
+    return edges_;
+  }
+
+  const std::list<ExecNode> &Nodes() const {
+    return nodes_;
+  }
+
+  const std::list<ExecEdge> &Edges() const {
+    return edges_;
+  }
+
+  ExecNode *Node(std::string_view name) {
+    auto it = name2node_.find(name);
+    if (it == name2node_.end())
+      return nullptr;
+    return it->second;
+  }
+
+  const ExecNode *Node(std::string_view name) const {
+    auto it = name2node_.find(name);
+    if (it == name2node_.end())
+      return nullptr;
+    return it->second;
+  }
+
+  template <typename... Args>
+  ExecNode *AddNode(Args &&...args) {
+    Invalidate();
+    ExecNode *node = &nodes_.emplace_back(std::forward<Args>(args)...);
+    if (!node->instance_name.empty()) {
+      if (!name2node_.emplace(node->instance_name, node).second) {
+        nodes_.pop_back();
+        throw std::invalid_argument(
+            make_string("Duplicate node name: \"", node->instance_name, "\""));
+      }
+    }
+    return node;
+  }
+
+  ExecNode *AddOutputNode() {
+    Invalidate();
+    ExecNode *node = &nodes_.emplace_back(PipelineOutputTag());
+    if (!node->instance_name.empty()) {
```

Comment:
No - there can be only one output node and `Validate` checks that.
It doesn't have an instance name, because the sole action that an output node does is to collect the results and repack the into the output workspace.

Reply:
This is why I am asking about this part of code, as it seems like a copy-paste from above for a case that shouldn't happen. I am not sure if it deserves earlier check here that prevents adding more, just noting the fact.